var/spack/repos/builtin/packages/quantum-espresso/package.py

```diff
@@ -60,30 +73,86 @@ class QuantumEspresso(Package):
         msg='elpa is a parallel library and needs MPI support'
     )
 
+    # HDF5 support introduced in 6.1 but requires MPI, develop
+    # branch and future releases will support serial HDF5
+    conflicts('+hdf5', when='@:6.0.0')
+    conflicts(
+        '+hdf5',
+        when='~mpi@6.1.0:6.3',
+        msg='HDF5 support only available with MPI for QE 6.1:6.3'
+    )
+
     # Elpa is formally supported by @:5.4.0, but QE configure searches
     # for it in the wrong folders (or tries to download it within
     # the build directory). Instead of patching Elpa to provide the
     # folder QE expects as a link, we issue a conflict here.
     conflicts('+elpa', when='@:5.4.0')
-    conflicts('+hdf5', when='@:5.4.0')
 
     # Spurious problems running in parallel the Makefile
     # generated by the configure
     parallel = False
 
+    # QE upstream patches
+    # QE 6.3 requires multiple patches to fix MKL detection
+    # There may still be problems on Mac with MKL detection
+    patch_url = 'https://gitlab.com/QEF/q-e/commit/0796e1b7c55c9361ecb6515a0979280e78865e36.diff'
+    patch_checksum = 'bc8c5b8523156cee002d97dab42a5976dffae20605da485a427b902a236d7e6b'
+    patch(patch_url, sha256=patch_checksum, when='@6.3')
+
+    # QE 6.3 `make install` broken and a patch must be applied
+    patch_url = 'https://gitlab.com/QEF/q-e/commit/88e6558646dbbcfcafa5f3fa758217f6062ab91c.diff'
+    patch_checksum = 'b776890d008e16cca28c31299c62f47de0ba606b900b17cbc27c041f45e564ca'
+    patch(patch_url, sha256=patch_checksum, when='@6.3')
```

Comment:
Just a heads up, these patches may break if and when a version 6.3.1 is released. See #8320, #8432, #8957, #9281, #9307, and #9330. To future proof this, you can hack it like so:
```suggestion
    patch(patch_url, sha256=patch_checksum, when='@6.3:6.3.0')
```
as was done in #9330.